### DISCUSSION TITLE
Managing / Storing Module Data

### DISCUSSION THREAD

**Discussion by Sharafudeen** [oqtane/oqtane.framework#3312] [General] [answered]
Hi,

I needs to build a module & store its data in a database table. I wanted to use that module in all pages (across all sites). Is it possible in Oqtane?

**Comment by sbwalker** [comment, marked as answer]
Yes it is possible. However when you say that you want to use it across all sites, the behavior is going to depend on your multi-tenant configuration. If you are using shared tenancy then all of your sites exist in the same database. So you can create a module which does not use a ModuleId or SiteId for scoping - the data will be global across the entire tenant (all sites). However if you are using isolated tenancy then each site has its own database... and you would want to use the "master" tenant's database to store information.

**Comment by Sharafudeen** [comment]
@sbwalker I can able to add OrganizationUnit module (with centralized data) in all pages in a root site. However I could not able to add the same OrganizationUnit module in the sub-site (without moduleId / siteId) which is exists in the same tenant / database. Please find the error & module records in the db screenshots.

**Issue Resolved:** While adding the module to the page, I've skipped the module title as "blank" then the issue has been resolved.

**_Also, I wanted to manage all admin modules and its data in centralized datastore in the root site instead of duplicating module data for each site._**

**Error:**
![image](https://github.com/oqtane/oqtane.framework/assets/42769344/9b2417e4-a116-4fe7-aca9-07a9f9da1e7f)

**DB Record:**
![image](https://github.com/oqtane/oqtane.framework/assets/42769344/a2e9f93d-7111-4882-b7cb-7b9021d98dff)

**Comment by sbwalker** [comment]
@Sharafudeen please explain what you mean by "While adding the module to the page, I've skipped the module title as "blank" then the issue has been resolved" - if the module title is left blank, the framework is supposed to use the module's name by default.

**Reply by Sharafudeen** [reply]
If module title in the sub-site is same as root site then the issue is raised. Means, couldn’t able to add the module in the sub-site.

**Reply by sbwalker** [reply]
A module title does not need to be unique. Can you please explain the exact steps to reproduce this problem. Use the Html/Text as the example module.

**Reply by Sharafudeen** [reply]
@sbwalker I’ve created the module as below steps.

1) Create a Oqtane module called OrganizationUnit
2) Removed ModuleId from OrganizationUnit
3) Build as a package and installed in Oqtane 4.0.5
4) Add OrganizationUnit module with module title as “Organization Unit” in the “Private” page in the root site
5) Add same OrganizationUnit module with module title as “Organization Unit” in the “Private” page in the sub site but it is not adding and throws an error.
6) If OrganizationUnit module with module title left as blank then it is adding to the page

**Reply by sbwalker** [reply]
@Sharafudeen what do you mean by "Removed ModuleId from OrganizationUnit"? Do you mean that you removed the property from the model?

**Reply by Sharafudeen** [reply]
Yes I’ve removed the ModuleId property from the model (as well as Client, Server and Shared projects).

**Reply by sbwalker** [reply]
@Sharafudeen I have created a custom module based on the default template, compiled the code, and tested adding the module to a parent and child site - and it works fine. So this would lead me to conclude that the problem is related to how you customized the code (ie. removing the ModuleId). It is impossible for me to diagnose this further.

**Reply by leigh-pointer** [reply]
Why?

**Reply by leigh-pointer** [reply]
@Sharafudeen the id's you mention are key, not only that they sound keybso why remove them! Why?

**Reply by sbwalker** [reply]
@Sharafudeen I wrote a blog about Data Scoping to provide some more insight into this area https://www.oqtane.org/blog/!/72/data-scoping. The ModuleId in the default module template is for data scoping. However if you do not need to scope your data at the module level I would suggest scoping it at the Site level instead, so that it still supports multi-tenancy.

**Reply by Sharafudeen** [reply]
@sbwalker Yes I’ve read that. Thanks for sharing and it is very helpful. I suggest to record (about the framework core) these kind of articles in special category under the documentation section.

**Comment by sbwalker** [comment]
@Sharafudeen your comment - "Also, I wanted to manage all admin modules and its data in centralized datastore in the root site instead of duplicating module data for each site"... it sounds like you do not want to use multi-tenancy as the whole idea behind multi-tenancy is that you separate the data for each site so that they can be managed independently.

**Reply by Sharafudeen** [reply]
I mean that, some of the admin modules data can be centralized data store. Let’s say, users can be shared to all the sites.

**Reply by sbwalker** [reply]
As I mentioned, the whole purpose wof multi-tenancy is that you separate the data for each site so that they can be managed independently. The one exception  in Oqtane is User information - as it can be shared across the same database. So if you have 2 sites which are stored in the same database then the users can be shared across those 2 sites. However ALL other types of data (ie. roles, pages, etc... ) are maintained independently for each site. This is fundamental to the architecture and is not configurable.

**Reply by Sharafudeen** [reply]
@sbwalker Thanks for the clarification.
